### ISSUE TITLE
ARM Support

### ISSUE THREAD

**Issue by t-0-m-1-3** [BishopFox/sliver#184]
**Describe the bug**
Trying to build sliver using the instructions for compiling from source, i'm unable to finish the build past the 18/23rd step. I have built it on Ubuntu and ParrotOS, I'm trying to pop it on a RP3

**To Reproduce**
1. clone and cd into repo
2. `docker build -t sliver .` **or** `./build.py` 
3. output below
```bash
Step 18/23 : RUN make static-linux && cp -vv sliver-server /opt/sliver-server
 ---> Running in f3069e268c83
printf "package version\n\nconst GitVersion = \"\"\n" > ./client/version/version.go
packr clean
rm -f ./protobuf/client/*.pb.go
rm -f ./protobuf/sliver/*.pb.go
rm -f sliver-client sliver-server *.exe
printf "package version\n\nconst GitVersion = \"%s\"\n" 9727b5491246390120397a215d8b0162a21610db > ./client/version/version.go
printf "const GitDirty = \"%s\"\n" Dirty >> ./client/version/version.go
protoc -I protobuf/ protobuf/sliver/sliver.proto --go_out=protobuf/
/usr/local/bin/protoc: 1: /usr/local/bin/protoc: Syntax error: word unexpected (expecting ")")
make: *** [Makefile:67: pb] Error 2
The command '/bin/sh -c make static-linux && cp -vv sliver-server /opt/sliver-server' returned a non-zero code: 2
```

**Desktop (please complete the following information):**
 - OS: Raspbian (Fails) 
- Parrot GNU/Linux 4 ( Works)
-  ubuntu 18.04 ( Works )

**Comment by moloch--** [comment]
Can you confirm branch/commit you're using? I'm unable to reproduce on the latest `master`

**Comment by rkervella** [comment]
From the error message and the OS versions, I suppose something's messed up for `protoc` on Raspbian.
Are all the tests run on a Raspberry Pi?

**Comment by moloch--** [comment]
Potentially an incompatibility with ARM? It should in theory be possible to compile/cross-compile Sliver to ARM but I've never tried.

**Comment by rkervella** [comment]
Same here, never tried to compile nor to execute on ARM.

**Comment by t-0-m-1-3** [comment]
> Can you confirm branch/commit you're using? I'm unable to reproduce on the latest `master`

It is the master branch; Everywhere else I build, it works fine, save my implant device

**Comment by moloch--** [comment]
I think I still have a couple RasperryPi's laying around let me see if I can help you troubleshoot.

**Comment by t-0-m-1-3** [comment]
Much appreciated
I owe you a :coffee: or a :beer:  if i see you around

**Comment by moloch--** [comment]
Okay bad news; I was able to cross-compile the server to ARM and get it to run on my RaspberryPi however BadgerDB, the internal database we use isn't compatible. We may be moving away from Badger in the future though, so it's possible we can get it to run after we move to something else.

**Comment by moloch--** [comment]
You can [see here](https://github.com/dgraph-io/badger/issues/246) for more information, it may be possible to tweak/hack around some of the database code to get it working.

To cross compile, on a desktop (x86/amd64) Linux machine follow the normal steps as described in the wiki but _before_ you run `make linux` run (you may need to tweak these depending on which version of the Raspberry Pi you have):

```
export GOARCH=arm
export GOARM=5
```

**Comment by t-0-m-1-3** [comment]
@moloch-- Much appreciated; I'm forking so I can try to carve out time to help; 

:beers: on me

**Comment by moloch--** [comment]
I should also warn you that the `generate` command uses an embedded version of Go (this what the `go-assets.sh` script does), so you if you want that command to work you'll need to cross-compile go itself too (which isn't hard, there's instructions on how to do this elsewhere). However, when using symbol obfuscation in `generate` we basically parse and refactor the implant code on the fly and this consumes a decent amount of CPU/memory, which likely will result in very long `generate` command run times. 

However, if you're just trying to catch shells you don't 100% need the `generate` command working, any implant with a matching TLS certificate can talk to the server regardless if it's ARM/x86.

**Comment by moloch--** [comment]
Just ordered an ARM-based MacBook, so I'll be working on first-class ARM support soon :D

**Comment by moloch--** [comment]
Fixed in v1.3.x
